### PR TITLE
test: restore supernodes after CC failure

### PR DESCRIPTION
**Pull request by Victor-C-Zhang**
- test correctness of post-failure supernode restore
- fix bug where DSU size array is not properly maintained

### CODE REVIEW

File: test/graph_test.cpp [hunk 1]
```diff
@@ -66,6 +67,42 @@ TEST_P(GraphTest, IFconnectedComponentsAlgRunTHENupdateLocked) {
...
+  while (m--) {
+    in >> a >> b;
+    g.update({{a, b}, INSERT});
+  }
+  g.set_verifier(std::make_unique<FileGraphVerifier>(curr_dir + "/res/multiples_graph_1024.txt"));
+  g.should_fail_CC();
+
+  // flush to make sure copy supernodes is consistent with graph supernodes
+  g.bf->force_flush();
+  GraphWorker::pause_workers();
```
Comment (Victor-C-Zhang) [inline test/graph_test.cpp:91]: let it be known I spent the entire day debugging the graph functions when this was the real reason the test was failing.

File: include/test/write_configuration.h [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #include <fstream>
 
-static void write_configuration(bool use_tree, bool backup_in_mem=false, int groups=1, int g_size=1) {
+static void write_configuration(bool use_tree, bool backup_in_mem = false, int
+groups = 1, int g_size = 1) {
```
Comment (etwest) [inline include/test/write_configuration.h]: Style nit: space before groups on line 4

Comment (etwest) [inline include/test/write_configuration.h:4]: Style nit: space before groups on line 4

File: include/graph.h [hunk 3]
```diff
@@ -127,6 +130,10 @@ class Graph {
   void set_verifier(std::unique_ptr<GraphVerifier> verifier) {
     this->verifier = std::move(verifier);
   }
+
+  // to induce a failure mid-CC
+  bool fail_round_2 = false;
```
Comment (etwest) [inline include/graph.h:135]: We should have these variables and all the other related code ifdef'd behind VERIFY_SAMPLES

Comment (etwest) [inline include/graph.h:135]: We should have these variables and all the other related code ifdef'd behind VERIFY_SAMPLES

File: src/graph.cpp [hunk 4]
```diff
@@ -22,10 +22,12 @@ Graph::Graph(node_id_t num_nodes): num_nodes(num_nodes) {
   representatives = new std::set<node_id_t>();
   supernodes = new Supernode*[num_nodes];
   parent = new node_id_t[num_nodes];
+  size = new node_id_t[num_nodes];
```
Comment (etwest) [inline src/graph.cpp:25]: Good change

Comment (etwest) [inline src/graph.cpp:25]: Good change

File: src/graph.cpp [hunk 5]
```diff
@@ -207,8 +210,9 @@ inline void Graph::merge_supernodes(Supernode** copy_supernodes, std::vector<nod
     // OMP requires a traditional for-loop to work
     node_id_t a = new_reps[i];
     try {
-      if (make_copy && copy_in_mem) // make a copy of a
+      if (make_copy && copy_in_mem) { // make a copy of a
```
Comment (etwest) [inline src/graph.cpp:213]: Why?

Comment (etwest) [inline src/graph.cpp:213]: Why?

File: test/graph_test.cpp (hunk 1 above)
Comment (etwest) [reply test/graph_test.cpp:91]: Oh no :D

Comment (etwest) [reply test/graph_test.cpp:91]: Oh no :D

File: src/graph.cpp (hunk 5 above)
Comment (Victor-C-Zhang) [reply src/graph.cpp:213]: mostly for clarity

File: include/graph.h (hunk 3 above)
Comment (Victor-C-Zhang) [reply include/graph.h:135]: they are

Comment (etwest) [reply include/graph.h:135]: Oh, awkward. 👍

**Review by etwest — APPROVED** [review]
Looks good

**Pull request merged**